framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java

```diff
@@ -267,97 +268,176 @@ protected boolean areEqualInHierarchy(
+            boolean result = isContainedByBoundType(inside, new BoundType(outside), canBeCovariant);
+            typeargVisitHistory.add(inside, outside, currentTop, result);
+            return result;
         }
-        Boolean previousResult = typeargVisitHistory.result(inside, outside, currentTop);
-        if (previousResult != null) {
-            return previousResult;
+        if (canBeCovariant) {
+            return isSubtype(inside, outside, currentTop);
         }
-
-        typeargVisitHistory.add(inside, outside, currentTop, true);
-        boolean result =
-                isContainedWildcard(
-                        inside, outside, outsideUpperBound, outsideLowerBound, canBeCovariant);
-        typeargVisitHistory.add(inside, outside, currentTop, result);
-        return result;
+        return areEqualInHierarchy(inside, outside);
     }
 
-    private boolean isContainedWildcard(
-            AnnotatedTypeMirror inside,
-            AnnotatedTypeMirror outside,
-            AnnotatedTypeMirror outsideUpperBound,
-            AnnotatedTypeMirror outsideLowerBound,
-            boolean canBeCovariant) {
-
-        if (inside.equals(outside)) {
-            // If they are equal, outside always contains inside.
-            return true;
-        }
-
-        if (inside.getKind() == TypeKind.WILDCARD) {
-            outsideUpperBound =
-                    checker.getTypeFactory()
-                            .widenToUpperBound(outsideUpperBound, (AnnotatedWildcardType) inside);
-        }
-        while (outsideUpperBound.getKind() == TypeKind.WILDCARD) {
-            if (ignoreUninferredTypeArgument(outsideUpperBound)) {
-                return true;
+    /**
+     * Returns true if {@code outside} contains {@code inside}, that is, if the set of types denoted
+     * by {@code outside} is a superset of or equal to the set of types denoted by {@code inside}.
+     *
+     * <p>A declared type is considered a supertype of another declared type only if all of the type
+     * arguments of the declared type "contain" the corresponding type arguments of the subtype.
+     * Containment is described in <a
+     * href="https://docs.oracle.com/javase/specs/jls/se11/html/jls-4.html#jls-4.5.1">JLS section
+     * 4.5.1 "Type Arguments of Parameterized Types"</a>.
+     *
+     * @param inside a possibly-contained type
+     * @param outside a possibly-containing type
+     * @param canBeCovariant whether or not type arguments are allowed to be covariant
+     * @return true if inside is contained by outside, or if canBeCovariant == true and {@code
+     *     inside <: outside}
+     */
+    protected boolean isContainedByBoundType(
+            AnnotatedTypeMirror inside, BoundType outside, boolean canBeCovariant) {
+        if (BoundType.isBoundType(inside)) {
+            BoundType insideBoundType = new BoundType(inside);
+            if (!insideBoundType.hasSuperBound && !outside.hasSuperBound) {
```

Comment:
I don't understand the logic of the next 7 lines.

One question is why there is a special case that depends on `hasSuperBound`.  I think that the lower bound will be the null type, and I would expect all the tests to work properly for such a type.

It would also be helpful to state the algorithm you are following.

I think it's something like this:
```
If Outside is not a wildcard:
  require equality
  return

if Inside is not a wildcard:
  Inside must be one of Outside's bounds
  return

let Outside = ? super Lo extends Uo
let Inside = ? super Li extends Ui
require Ui <: Uo
        Lo <: Li
```

(which is itself just a simplification of the first 5 bullet points in the containment algorithm of JLS 4.5.1) plus special cases for contravariance that I have not written down.  Is the algorithm as I stated it correct or incorrect?